### PR TITLE
Add hard dependency on setuptools for Py3.12 support

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** The PennyLane device loader makes use of `pkg_resources`, which is included as part of `setuptools`, and is default installed into a virtual-environment with all PennyLane-supported Python versions <=3.11. Python 3.12 removed the default installation of setuptools (https://github.com/python/cpython/issues/95299) and so the `pkg_utils` import will fail, and the device loader will not work. This fix ensures `setuptools` and `pkg_resources` are hard dependencies of PennyLane to rectify the above.

**Description of the Change:** Create hard dependency on setuptools package.

**Benefits:** Fixes failed imports on Python 3.12

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "pennylane-lightning>=0.33",
     "requests",
     "typing_extensions",
+    "setuptools",
```
Comment (josh146) [inline setup.py:36]: @mlxd will this... work? Don't you need `setuptools` to run the logic in `setup.py`?

Comment (josh146) [reply setup.py:36]: E.g., this file will fail at line 17
```python
from setuptools import setup, find_packages
```

Comment (josh146) [reply setup.py:36]: iirc this is the catch-22 that `pyproject.toml` was trying to solve -- letting you specify build dependencies independent of the build system

Comment (mlxd) [reply setup.py:36]: The issue is for PennyLane at runtime in a virtual-env --- the module is already available for building the wheels, as the wheel package depends on it (and cibuildwheel will handle that). We need `pkg_resources` for the device loader at runtime, so we need it in any environment the wheel is installed onto.

Comment (josh146) [reply setup.py:36]: got it! Maybe worth including `pkg_resources` here then instead of `setuptools` to be more specific?

Comment (mlxd) [reply setup.py:36]: Works fine in a venv with the above addition.

**Comment by mlxd** [comment]
Closing in favour of https://github.com/PennyLaneAI/pennylane/pull/4783

**Pull request closed without merging**